presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1330,6 +1391,153 @@ public synchronized void progressChanged(ProgressEvent progressEvent)
+                long multiPartUploadMinPartSize)
+        {
+            STATS.uploadStarted();
+
+            this.s3 = s3;
+
+            int bufferSize = (int) multiPartUploadMinPartSize;
+            this.buf = new byte[bufferSize];
+
+            this.bucketName = bucketName;
+            this.key = key;
+            this.uploadId = uploadId;
+
+            this.uploadExecutor = uploadExecutor;
+
+            this.uploadFutures = new ArrayList<>();
+        }
+
+        @Override
+        public synchronized void write(int b)
+        {
+            if (count >= buf.length) {
+                flushBuffer(false);
+            }
+            buf[count++] = (byte) b;
+        }
+
+        @Override
+        public synchronized void flush()
+        {
+            flushBuffer(false);
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            flushBuffer(true);
+
+            if (!uploadFutures.isEmpty()) {
+                ListenableFuture<List<UploadPartResult>> partsFuture = allAsList(uploadFutures);
+                ListenableFuture<?> finishFuture = transform(partsFuture, this::finishUpload, uploadExecutor);
+
+                try {
+                    // Sync call to block presto from doing any other thing before the file is uploaded.
+                    finishFuture.get();
+                }
+                catch (InterruptedException e) {
+                    abortUpload(true);
+                    Thread.currentThread().interrupt();
+                    throw new InterruptedIOException();
+                }
+                catch (ExecutionException e) {
+                    abortUpload(true);
+                    throw new IOException(e.getCause());
+                }
+                catch (Exception e) {
+                    abortUpload(true);
+                    throw new IOException(e);
+                }
+            }
+            else {
+                abortUpload(false);
+            }
+        }
+
+        private void flushBuffer(boolean finished)
+        {
+            // The multipart upload API only accept the last part to be less than 5MB
+            if (count == buf.length || (finished && count > 0)) {
+                byte[] data = Arrays.copyOf(buf, count);
+                ListenableFuture<UploadPartResult> future = uploadExecutor.submit(
+                        () -> uploadPage(data));
+                uploadFutures.add(future);
+                count = 0;
+            }
+        }
+
+        private UploadPartResult uploadPage(byte[] data)
+        {
+            int partNumber = ++currentPartNumber;
+            UploadPartRequest uploadRequest = new UploadPartRequest()
+                    .withBucketName(bucketName)
+                    .withKey(key)
+                    .withUploadId(uploadId)
+                    .withPartNumber(partNumber)
+                    .withInputStream(new ByteArrayInputStream(data))
```

Comment:
We should add the MD5 so that S3 can verify the integrity of the upload
```java
.withMD5Digest(getMd5AsBase64(data))
```
```java
private static String getMd5AsBase64(byte[] data)
{
    @SuppressWarnings("deprecation")
    byte[] md5 = md5().hashBytes(data).asBytes();
    return Base64.getEncoder().encodeToString(md5);
}
```